Documenting Node and NodeGroup

leenlab2 opened the pull request:
Added documentation for the functions and methods in `Node.js` and `NodeGroup.js`, as well as comments explaining certain sections of code.

There were some functions that I did not fully understand, so I did not add proper documentation to them but instead placed a template of one that we could update in the future.

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
     var state = localStorage.getItem(this.props.JSON.id_);
+
+    /**
+     * Unselected states:
```
This is great documentation, but move it to the class documentation above.

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
     if (this.props.editMode) {
       state = ""; // TODO: define what editMode is
     } else if (state === null) {
+      // sets state to takeable if it has no parent nodes (prereqs)
```
Remove this comment. It is redundant with the description of the states.

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
+      // sets state to takeable if it has no parent nodes (prereqs)
       state = this.props.parents.length === 0 ? "takeable" : "inactive";
     }
+
```
Please revert all blank line changes. They aren't part of the scope of this pull request.

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
     this.state = {
       status: state,
-      selected: ["active", "overridden"].indexOf(state) >= 0
+      selected: ["active", "overridden"].indexOf(state) >= 0 // checks if state means its selected
```
Remove this comment. Instead, it's better to add a note in the class documentation above that `selected` is a state.

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
+   * @return {boolean}
+   */
   isSelected = () => {
+    // Hybrid nodes: the grey, smaller nodes representing another node farther way
```
Again, move this to the class documentation, as a separate part where you discuss what hybrid nodes are.

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
   }
 
+  /**
+   * Checks whether all prerequisite/ preceeding nodes for the current one are satisfied
```
Do not include spaces after `/`. Also, "preceeding" is spelled "preceding".

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
+     * @return {boolean}
+     */
     function isAllTrue(element) {
+      // base case: single node
```
Remove all inline comments in this function, and the one on line 73; they are just repeating what the code does, and in general we should not include such comments.

leenlab2 replied:
Should I remove the one next to var svg...?

david-yz-liu replied:
Oh, yes that one too

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
   updateNode = recursive => {
     var newState;
+
+    // if the prereqs are satisfied, the node is takeable. Once selected it becomes active.
```
Remove the inline comments in this function (same comment as above).

leenlab2 replied:
The one on line 107 was not added by me, should I keep it?

david-yz-liu replied:
Good question! That's an example of a comment which isn't adding much, but could be improved to explain the *why* of the condition. So why don't you change it to a comment like "We can stop without updating the children when the selected state of this node hasn't changed and the original state wasn't "missing"."

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -97,19 +143,23 @@ export default class Node extends React.Component {
   focusPrereqs = () => {
-    var svg = this.props.svg;
-    // Check if there are any missing prerequisites.
+    var svg = this.props.svg;  // contains information about the whole graph
```
Remove the inline comments

leenlab2 replied:
The comment on line 157 was also not added by me, should I improve or delete it?

david-yz-liu replied:
Hmm, you can try to improve that one!

leenlab2 replied:
I decided to improve it to answer "why",  like we did to the other one

david-yz-liu commented on js/components/graph/Node.js:
```diff
@@ -139,10 +192,17 @@ export default class Node extends React.Component {
   unfocusPrereqs = () => {
-    var svg = this.props.svg;
-    this.updateNode(false);
+    var svg = this.props.svg;  // contains information about the whole graph
```
Remove the inline comments

david-yz-liu commented on js/components/graph/NodeGroup.js:
```diff
@@ -2,15 +2,19 @@ import PropTypes from "prop-types";
 export default class NodeGroup extends React.Component {
+  /** Returns nodes to their original unselected state, with a status of "takeable" or "inactive" */
   reset = () => {
+    // Resets course nodes
```
Remove the inline comments

david-yz-liu commented on js/components/graph/NodeGroup.js:
```diff
@@ -43,17 +56,21 @@ export default class NodeGroup extends React.Component {
           var childs = [];
           var outEdges = [];
+
+          // build childs and outEdges
```
Remove this comment

david-yz-liu commented on js/components/graph/NodeGroup.js:
```diff
@@ -43,17 +56,21 @@ export default class NodeGroup extends React.Component {
         {this.props.hybridsJSON.map(entry => {
           var childs = [];
           var outEdges = [];
+
```
Remove the blank line changes here.

david-yz-liu reviewed (commented):
@leenlab2 a good start. I left a few different comments for you to address. My overall comment is that you should not be including inline comments to describe what code does, as the code already communicates this. (This is different from CS classes where you are explicitly learning the language. The standard for reading and understanding code is higher in a real-world project.)

A good rule of thumb for inline comments is that you should only include them to describe *why* a piece of code is written the way that it is, rather than *what* the code does.

leenlab2 commented:
I applied some of your comments with these new commits. That includes:

- Removed blank line changes
- Removed unnecessary inline comments
- Made the class description more detailed, including describing the states, the status messages, and the types of nodes

Let me know what else needs fixing!

david-yz-liu commented:
@leenlab2 great changes! You should leave the comments open so that I can see them when I review your changes.
